File: Algorithm_63_Unique Paths II.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid):
        m,n=len(obstacleGrid),len(obstacleGrid[0])
        mat=[[1 for j in range(n)] for i in range(m)]
        for i in range(m):
            if obstacleGrid[i][0]==1:
                for j in range(i,m): mat[j][0]=0
                break
        for j in range(n):
            if obstacleGrid[0][j]==1:
                for i in range(j,n): mat[0][i]=0
                break
        for i in range(1,m):
            for j in range(1,n):
                if obstacleGrid[i][j]==1:
                    mat[i][j]=0
                elif obstacleGrid[i-1][j]==0 and obstacleGrid[i][j-1]==0:
                    mat[i][j]=mat[i-1][j]+mat[i][j-1]
                elif obstacleGrid[i-1][j]==1:
                    mat[i][j]=mat[i][j-1]
                else:
                    mat[i][j]=mat[i-1][j]
        return mat[m-1][n-1]

File: Algorithm_63_Unique Paths II.java
// 思路：动态规划。dp[i][j]记录走到[i][j]格子上的所有路径数
// 注意初始状态为dp[0][0]，要判断obstacleGrid[0][0]是否为1
// dp[i][j] = dp[i - 1][j] + dp[i][j - 1]，注意i == 0和j == 0
// 可以优化到一维数组

class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        if (obstacleGrid == null || obstacleGrid.length == 0 || obstacleGrid[0] == null || obstacleGrid[0].length == 0) {
            return 0;
        }
        int[][] dp = new int[obstacleGrid.length][obstacleGrid[0].length];
        dp[0][0] = obstacleGrid[0][0] == 0 ? 1 : 0; // don't forget to judge whether obstacleGrid[0][0] == 1
        for (int i = 0; i < obstacleGrid.length; i++) {
            for (int j = 0; j < obstacleGrid[0].length; j++) {
                if (obstacleGrid[i][j] == 0) { // available
                    dp[i][j] += j > 0 ? dp[i][j - 1] : 0;
                    dp[i][j] += i > 0 ? dp[i - 1][j] : 0;
                }
            }
        }
        return dp[dp.length - 1][dp[0].length - 1];
    }
}